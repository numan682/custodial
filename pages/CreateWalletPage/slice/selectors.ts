import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { initialState } from '.'

const selectSlice = (state: RootState) => state.createWallet || initialState

export const selectCreateWallet = createSelector([selectSlice], state => state)
export const selectMnemonic = createSelector([selectCreateWallet], wallet => wallet.mnemonic)
export const selectCheckbox = createSelector([selectCreateWallet], wallet => wallet.checkbox)
