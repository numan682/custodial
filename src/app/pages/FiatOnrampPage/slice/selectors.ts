import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { initialState } from '.'

const selectSlice = (state: RootState) => state.fiatOnramp || initialState

export const selectFiatOnramp = createSelector([selectSlice], state => state)
export const selectThirdPartyAcknowledged = createSelector(
  [selectFiatOnramp],
  wallet => wallet.thirdPartyAcknowledged,
)
