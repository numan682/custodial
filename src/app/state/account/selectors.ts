import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { initialState } from '.'

const selectSlice = (state: RootState) => state.account || initialState

export const selectAccount = createSelector([selectSlice], state => state)
export const selectTransactions = createSelector([selectAccount], account => account.transactions)
export const selectTransactionsError = createSelector([selectAccount], account => account.transactionsError)
export const selectAccountAddress = createSelector([selectAccount], account => account.address)
export const selectAccountAvailableBalance = createSelector([selectAccount], account => account.available)
export const selectAccountIsLoading = createSelector([selectAccount], account => account.loading)
export const selectAccountAllowances = createSelector([selectAccount], account => account.allowances)
