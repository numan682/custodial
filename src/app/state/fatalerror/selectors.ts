import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { initialState } from '.'

const selectSlice = (state: RootState) => state.fatalError || initialState

export const selectFatalError = createSelector([selectSlice], state => state.error)
