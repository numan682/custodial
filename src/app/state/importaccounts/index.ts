import { PayloadAction } from '@reduxjs/toolkit'
import { ErrorPayload } from 'types/errors'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { ImportAccountsListAccount, ImportAccountsState, ImportAccountsStep } from './types'

export const initialState: ImportAccountsState = {
  accounts: [],
  showAccountsSelectionModal: false,
  accountsSelectionPageNumber: 0,
  step: ImportAccountsStep.Idle,
}

const slice = createSlice({
  name: 'importAccounts',
  initialState,
  reducers: {
    clear(state) {
      state.accounts = []
      state.error = undefined
      state.step = ImportAccountsStep.Idle
      state.showAccountsSelectionModal = false
    },
    enumerateAccountsFromLedger(state) {
      state.accounts = []
      state.accountsSelectionPageNumber = 0
      state.showAccountsSelectionModal = true
      state.step = ImportAccountsStep.Idle
    },
    enumerateMoreAccountsFromLedger(state) {
      state.step = ImportAccountsStep.Idle
    },
    enumerateAccountsFromMnemonic(state, _action: PayloadAction<string>) {
      state.step = ImportAccountsStep.Idle
      state.accounts = []
      state.accountsSelectionPageNumber = 0
      state.showAccountsSelectionModal = true
    },
    toggleAccount(state, action: PayloadAction<string>) {
      const account = state.accounts.find(a => a.address === action.payload)!
      account.selected = !account.selected
    },
    accountGenerated(state, action: PayloadAction<ImportAccountsListAccount>) {
      state.accounts.push(action.payload)
    },
    accountsListed(state, action: PayloadAction<ImportAccountsListAccount[]>) {
      state.accounts = action.payload
    },
    updateAccountBalance(
      state,
      action: PayloadAction<Pick<ImportAccountsListAccount, 'address' | 'balance'>>,
    ) {
      const { address, balance } = action.payload
      const account = state.accounts.find(a => a.address === address)
      if (account) account.balance = balance
    },
    setPage(state, action: PayloadAction<number>) {
      state.accountsSelectionPageNumber = action.payload
    },
    setStep(state, action: PayloadAction<ImportAccountsStep>) {
      state.step = action.payload
    },
    operationFailed(state, action: PayloadAction<ErrorPayload>) {
      state.error = action.payload
      state.step = ImportAccountsStep.Idle
    },
  },
})

export const { actions: importAccountsActions } = slice

export default slice.reducer
