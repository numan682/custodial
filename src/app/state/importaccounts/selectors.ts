import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { initialState } from '.'
import { accountsPerPage } from './saga'

const selectSlice = (state: RootState) => state.importAccounts || initialState

export const selectError = createSelector([selectSlice], state => state.error)
export const selectImportAccounts = createSelector([selectSlice], state => state)
export const selectImportAccountsFullList = createSelector([selectSlice], state => state.accounts)
export const selectShowAccountsSelectionModal = createSelector(
  [selectSlice],
  state => state.showAccountsSelectionModal,
)
export const selectImportAccountsPageNumber = createSelector(
  [selectSlice],
  state => state.accountsSelectionPageNumber,
)
export const selectImportAccountsOnCurrentPage = createSelector(
  [selectImportAccountsFullList, selectImportAccountsPageNumber],
  (fullList, pageNumber) => fullList.slice(pageNumber * accountsPerPage, (pageNumber + 1) * accountsPerPage),
)
export const selectImportAccountHasMissingBalances = createSelector(
  [selectImportAccountsOnCurrentPage],
  list => list.some(a => !a.balance),
)
export const selectSelectedAccounts = createSelector([selectImportAccountsFullList], state =>
  state.filter(a => a.selected),
)
