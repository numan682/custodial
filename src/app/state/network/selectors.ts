import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { initialState } from '.'

const selectSlice = (state: RootState) => state.network || initialState

export const selectSelectedNetwork = createSelector([selectSlice], state => state.selectedNetwork)
export const selectTicker = createSelector([selectSlice], state => state.ticker)
export const selectChainContext = createSelector([selectSlice], state => state.chainContext)
export const selectEpoch = createSelector([selectSlice], state => state.epoch)
export const selectMinStaking = createSelector([selectSlice], state => state.minimumStakingAmount)
