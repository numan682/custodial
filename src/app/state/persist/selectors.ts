import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { getInitialState } from '.'

export const selectSlice = (state: RootState) => state.persist || getInitialState()

export const selectLoading = createSelector([selectSlice], state => state.loading)
export const selectEnteredWrongPassword = createSelector([selectSlice], state => state.enteredWrongPassword)
export const selectNeedsPassword = createSelector(
  [selectSlice],
  state => state.hasPersistedProfiles && !state.stringifiedEncryptionKey,
)
export const selectSkipUnlockingOnInit = createSelector(
  [selectSlice],
  state => !state.hasPersistedProfiles && !state.stringifiedEncryptionKey,
)
export const selectHasPersistedProfiles = createSelector([selectSlice], state => state.hasPersistedProfiles)
export const selectIsPersistenceUnsupported = createSelector(
  [selectSlice],
  state => state.isPersistenceUnsupported,
)
