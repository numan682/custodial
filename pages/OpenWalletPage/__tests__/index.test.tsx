import React from 'react'
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { useSelector } from 'react-redux'
import { when } from 'jest-when'
import { selectShowAccountsSelectionModal } from 'app/state/importaccounts/selectors'
import { canAccessNavigatorUsb } from 'app/lib/ledger'
import { SelectOpenMethod } from '..'

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}))
jest.mock('app/lib/ledger')

const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

const renderComponent = (webExtensionLedgerAccess?: () => void) =>
  render(
    <MemoryRouter>
      <SelectOpenMethod webExtensionLedgerAccess={webExtensionLedgerAccess} />
    </MemoryRouter>,
  )

describe('<SelectOpenMethod />', () => {
  beforeEach(() => {
    jest.mocked(canAccessNavigatorUsb).mockResolvedValue(false)
  })

  it('should render component', () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
    expect(screen.getByText('errors.usbTransportNotSupported')).toBeInTheDocument()
  })

  it('should render component with an access button', () => {
    renderComponent(() => {})

    expect(screen.queryByText('openWallet.method.ledger')).not.toBeInTheDocument()
    expect(screen.getByText('ledger.extension.grantAccess')).toBeInTheDocument()
  })

  it('should redirect user to ledger page', () => {
    when(useSelector as any)
      .calledWith(selectShowAccountsSelectionModal)
      .mockReturnValue(true)

    renderComponent(() => {})

    expect(mockNavigate).toHaveBeenCalledWith('/open-wallet/ledger')
  })

  it('should render variant with web usb support', async () => {
    jest.mocked(canAccessNavigatorUsb).mockResolvedValue(true)

    const { rerender } = renderComponent()
    rerender(
      <MemoryRouter>
        <SelectOpenMethod />
      </MemoryRouter>,
    )

    await waitForElementToBeRemoved(() => screen.queryByText('errors.usbTransportNotSupported'))
    expect(screen.getByRole('button', { name: 'openWallet.method.ledger' })).not.toBeDisabled()
  })
})
