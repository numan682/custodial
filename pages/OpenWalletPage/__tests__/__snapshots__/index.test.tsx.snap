// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectOpenMethod /> should render component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin: 12px;
  background-color: #FFFFFF;
  color: #444444;
  border: solid 1px background-front-border;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 24px;
  border-radius: 5px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c4 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 24px;
}

.c6 {
  font-size: 14px;
  line-height: 20px;
  text-align: center;
}

.c3 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c3:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c3:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus > circle,
.c3:focus > ellipse,
.c3:focus > line,
.c3:focus > path,
.c3:focus > polygon,
.c3:focus > polyline,
.c3:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus::-moz-focus-inner {
  border: 0;
}

.c3:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c3:focus:not(:focus-visible) > circle,.c3:focus:not(:focus-visible) > ellipse,
.c3:focus:not(:focus-visible) > line,.c3:focus:not(:focus-visible) > path,
.c3:focus:not(:focus-visible) > polygon,.c3:focus:not(:focus-visible) > polyline,
.c3:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c3:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c5 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  opacity: 0.3;
  cursor: default;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c5:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c5:focus > circle,
.c5:focus > ellipse,
.c5:focus > line,
.c5:focus > path,
.c5:focus > polygon,
.c5:focus > polyline,
.c5:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c5:focus::-moz-focus-inner {
  border: 0;
}

.c5:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c5:focus:not(:focus-visible) > circle,.c5:focus:not(:focus-visible) > ellipse,
.c5:focus:not(:focus-visible) > line,.c5:focus:not(:focus-visible) > path,
.c5:focus:not(:focus-visible) > polygon,.c5:focus:not(:focus-visible) > polyline,
.c5:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c5:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c1 {
  margin-top: 0px;
  font-size: 34px;
  line-height: 40px;
  max-width: 816px;
  font-weight: 600;
  overflow-wrap: break-word;
}

@media only screen and (max-width:768px) {
  .c0 {
    margin: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c0 {
    border: solid 1px background-front-border;
  }
}

@media only screen and (max-width:768px) {
  .c0 {
    padding: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    margin-top: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-flex-basis: auto;
    -ms-flex-preferred-size: auto;
    flex-basis: auto;
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    -webkit-align-items: stretch;
    -webkit-box-align: stretch;
    -ms-flex-align: stretch;
    align-items: stretch;
  }
}

@media only screen and (max-width:768px) {
  .c4 {
    width: auto;
    height: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c1 {
    margin-top: 0px;
  }
}

@media only screen and (max-width:768px) {
  .c1 {
    font-size: 26px;
    line-height: 32px;
    max-width: 624px;
  }
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      openWallet.header
    </h1>
    <div
      class="c2"
    >
      <span>
        <a
          aria-label="openWallet.method.mnemonic"
          href="/mnemonic"
        >
          <button
            class="c3"
            tabindex="-1"
            type="button"
          >
            openWallet.method.mnemonic
          </button>
        </a>
      </span>
      <div
        class="c4"
      />
      <span>
        <a
          aria-label="openWallet.method.privateKey"
          href="/private-key"
        >
          <button
            class="c3"
            tabindex="-1"
            type="button"
          >
            openWallet.method.privateKey
          </button>
        </a>
      </span>
      <div
        class="c4"
      />
      <div>
        <div>
          <span>
            <a
              aria-label="openWallet.method.ledger"
              href="/ledger"
            >
              <button
                class="c5"
                disabled=""
                tabindex="-1"
                type="button"
              >
                openWallet.method.ledger
              </button>
            </a>
          </span>
        </div>
        <span
          class="c6"
        >
          errors.usbTransportNotSupported
        </span>
      </div>
    </div>
    <div
      class="c2"
      style="white-space: pre-wrap; display: inline;"
    />
  </div>
</div>
`;
